### PR TITLE
Support selection of exceptions that can be retryed or not

### PR DESCRIPTION
**Pull request by doumdoum**
Support selection of exceptions that can be retryed or that should not be retryed.

### CODE REVIEW

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetryInterceptor.groovy [hunk 1]
```diff
@@ -30,6 +40,9 @@ class RetryInterceptor implements IMethodInterceptor {
             } catch (org.junit.AssumptionViolatedException e) {
                 throw e
             } catch (Throwable t) {
+                if( ! checkRetry.test(t)){
```
Comment (anotherchrisberry) [inline src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetryInterceptor.groovy]:
```suggestion
                if (!checkRetry.test(t)) {
```
minor formatting nit, just to keep it consistent with the rest of the code

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailure.groovy [hunk 2]
```diff
@@ -15,4 +15,7 @@ public @interface RetryOnFailure {
     int times() default 1;
     int delaySeconds() default 0;
 
+    Class<? extends Throwable>[] only() default [];
```
Comment (anotherchrisberry) [inline src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailure.groovy]: It's probably a little easier to remember what this field is if it's named `include`, so you have `include` and `exclude` as the options.

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy [hunk 3]
```diff
@@ -44,7 +46,25 @@ class RetrySpecExtension extends AbstractAnnotationDrivenExtension<RetryOnFailur
     }
     double getDelaySeconds(RetryOnFailure retries) {
         String defaultDelay = Double.toString(retries.delaySeconds())
-        return Double.parseDouble(System.getProperty("spck-retry.delaySeconds", defaultDelay))
+        return Double.parseDouble(System.getProperty("spock-retry.delaySeconds", defaultDelay))
```
Comment (anotherchrisberry) [inline src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy:49]: nice catch

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy [hunk 4]
```diff
@@ -44,7 +46,25 @@ class RetrySpecExtension extends AbstractAnnotationDrivenExtension<RetryOnFailur
...
+    }
+
+    Predicate<Throwable> getPredicate(final RetryOnFailure retries) {
```
Comment (anotherchrisberry) [inline src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy]: Maybe rename to `getExceptionChecks` or something more descriptive about what the `Predicate` is for?

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy [hunk 5]
```diff
@@ -44,7 +46,25 @@ class RetrySpecExtension extends AbstractAnnotationDrivenExtension<RetryOnFailur
...
+        final Predicate<Throwable> predOnly = Optional.ofNullable(retries.only())
```
Comment (anotherchrisberry) [inline src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy]: same comment above about `Only` vs `Includes`

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy [hunk 6]
```diff
@@ -44,7 +46,25 @@ class RetrySpecExtension extends AbstractAnnotationDrivenExtension<RetryOnFailur
...
+                .map{l -> new HashSet<Throwable>(l)}
+                .map{c -> {t -> hasMatchingClass(t, c)}}
+                .orElseGet { -> {t -> true} }
+
+        final Predicate<Throwable> predExcludes = Optional.ofNullable(retries.exclude())
+                .map{l -> Arrays.asList(l)}
+                .filter{l -> ! l.isEmpty()}
+                .map{l -> new HashSet<Throwable>(l)}
+                .map{c -> {t -> ! hasMatchingClass(t, c)}}
+                .orElseGet { -> {t -> true} }
```
Comment (anotherchrisberry) [inline src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy:63]: Can this be simplified to:
```suggestion
        final Predicate<Throwable> predExcludes = Optional.ofNullable(retries.exclude())
                .map{l -> new HashSet<Throwable>(Arrays.asList(l))}
                .map{c -> {t -> !hasMatchingClass(t, c)}}
                .orElseGet { -> {t -> true} }
```
I don't think you necessarily need to `filter`, do you?

File: src/test/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailureOnlyExclude.groovy [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    void 'expect thrown on excluded exception'() {
+        when:
+        thrownTries++
+        throw new IllegalArgumentException('no problem')
+
+        then:
+        thrown(IllegalArgumentException)
+        thrownTries == 1
+    }
+}
```
Comment (anotherchrisberry) [inline src/test/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailureOnlyExclude.groovy]:
```suggestion
}

```
missing newline

File: src/test/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailureOnlyExclude.groovy [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package com.anotherchrisberry.spock.extensions.retry
```
Comment (anotherchrisberry) [inline src/test/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailureOnlyExclude.groovy]: I'd love to see one more test against `only`, where some other exception is thrown, just to verify that `only` will restrict the retry to the specified classes of exceptions.

**Comment by anotherchrisberry** [comment]
I really like this! Some minor formatting and naming things aside, this is great.

File: src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy (hunk 6 above)
Comment (doumdoum) [reply src/main/groovy/com/anotherchrisberry/spock/extensions/retry/RetrySpecExtension.groovy:63]: Yes. Method has matchingClass returns a relevant result only on non empty collections.
To simplify it i removed the HashSet stuff which is unnecessary.

File: src/test/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailureOnlyExclude.groovy (hunk 8 above)
Comment (doumdoum) [reply src/test/groovy/com/anotherchrisberry/spock/extensions/retry/RetryOnFailureOnlyExclude.groovy]: Done although the remainder of testsuite should be sufficient (no existing test has been broken).

**Comment by doumdoum** [comment]
Hi, 
Thx for your fast review.
Would be great to add CI in project.
FYI when i run tests within intellij it runs fine but when i run them from command line i got an error on: `com.anotherchrisberry.spock.extensions.retry.RetryOnFailureSubSubclassSpec > feature override FAILED
    java.lang.Exception at RetryOnFailureSubSubclassSpec.groovy:23
`
It fails also on master.

**Comment by anotherchrisberry** [comment]
Released in [0.6.5](https://bintray.com/anotherchrisberry/spock-retry/spock-retry/0.6.5) - thanks again!

**Pull request merged**